### DISCUSSION TITLE
Decouple SDK and engine releases and versions

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#3456] [Ideas]
## Overview

Each Dagger SDK is its own software project, distinct from, and with a dependency on, the Dagger Engine.

```mermaid
graph TD;

py["Python SDK"]
go["Go SDK"]
ts["Typescript SDK"]
js["Javascript SDK"]
classic["Dagger classic"]
engine["Dagger Engine"]

py & go & ts & js & classic --> engine
```

Since each SDK is developed independently, they should be versioned released independently as well.

This also applies to Dagger Classic, the compat layer for Dagger 0.2 CUE configurations.

## Questions

* What version schemes should we use?
* Should each sdk have its own repo?
* Assuming we have consensus on this, when should we execute this change?

cc @gerhard (master releaser) @aluzzardi (master architect) @mircubed (master planner)

**Comment by grouville** [comment]
Might be helpful/related: https://github.com/josh-project/josh (pretty sure Gerhard and Andrea already have it in mind)

**Comment by gerhard** [comment]
This is what I think:

**Q: What version schemes should we use for the SDKs?**
Semver, starting with `0.1.0` for the first release.

**Q: Should each SDK have its own repo?**
No. If we spread, the coordination work will increase by `O(n²)` [a.k.a. `O(my)`](https://twitter.com/ossia/status/1578061565965881344). Keeping all issues / PRs & discussions in a single repository keeps all context in a single place, with minimal synchronisation (reference the right ID). Having experienced first-hand, over the span of >5 years, the long-term (>10 years) trade-offs of multiple (>40) repositories, I would like us to avoid having to deal with those trade-offs. This is a rare situation where context doesn't make much difference.

All automation for various releases will live in a single repository - the current one.

Everyone can go to a single place to understand the state of everything public which is Dagger-related:
- engine
- docs
- SDKs

The above applies to metric scrapers too.

**Q: Assuming we have consensus on this, when should we execute this change?**
Before we produce the first SDK release. Since that is the Go SDK, I am very curious what @aluzzardi & @sipsma think about the above.

What do others think? cc @sipsma @aluzzardi @helderco @samalba @shykes

**Comment by helderco** [comment]
👍 for the monorepo.

I'd like to see everything SDK related under the same file tree, even the Go bits. So still make it modular and easy to break into a separate repo.

If you (as a company) want to create your own SDK, you should be able to copy that tree into a new empty repo that you own (e.g., `github.com/dagger/dagger/sdk/nodejs` → `github.com/acme/dagger-deno-sdk`).

**Comment by sipsma** [comment]
> Q: What version schemes should we use for the SDKs?

Agree with semver. 

One question is whether the Go SDK should be a separate go module from the rest of our codebase (right now it's not). To be explicit, I'm talking about a submodule, not splitting out to separate repo.

I don't currently see any advantage to doing this really. The only "public" part of our go codebase should be the Go SDK I think, so I don't see a need to version it and the rest of the code separately.
* However, that does imply that we should move most of our current code to an` internal/` package if we want to be prim and proper and follow all the right go conventions. I believe @aluzzardi mentioned this in a PR comment before if I'm not imagining things.

> Q: Should each SDK have its own repo?

Another +1 for monorepo. There was some discussion in the context of the nodejs sdk a while back here that came to the same conclusion: https://github.com/dagger/dagger/issues/3096#issue-1376428167

**Comment by shykes** [comment]
As @mircubed pointed out, we have a short term deadline in the form of the upcoming 0.3.0-alpha.1 release and, shortly after that, launching Go SDK in tech preview. Let's focus on the decisions we *need* to make before this deadline, and leave the rest for later.

Here are our options:

| | Option 1: coupled | Option 2: decoupled |
| -- |-- | -- |
| Versioning scheme for Go SDK | Coupled to engine | Semver |
| Version of Go SDK for tech preview next week | 0.3.0-alpha.1 | 0.1.0-alpha.1 |
| Version of engine used by Go SDK for tech preview next week | 0.3.0-alpha.1 | 0.3.0-alpha.1 |
| Reversible later | Yes (but couldn't rewind SDK versions: starting version would be engine version or higher) | No |
| SDK in same repo as engine? | Yes | Yes |
| Each SDK controls its engine dependency | No 👎  | No 👍 |
| SDK-specific improvement forces engine version bump | Yes 👎 | No 👍 |
| SDK-specific improvement forces version bump on other SDKs | Yes 👎 | No 👍 |
| Additional work to implement | No 👍 | Yes 👎 |

So far I'm seeing more arguments in favor of decoupling - the main downside being that we need to do it quickly (before 0.3.0-alpha.1 basically).

- Did I miss anything in this decision?
- Are there other decisions we need to make before 0.3.0-alpha.1?

Thanks!

**Comment by shykes** [comment]
This has been resolved. We are going for option 2: decoupled.
